### PR TITLE
Store idmap with metadata in source file.

### PR DESCRIPTION
**Pull request by Josef-Vonasek**
### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: common/syntax/specialization/shared/src/test/scala/org/enso/syntax/text/ParserTest.scala [hunk 1]
```diff
@@ -22,7 +22,7 @@ class ParserTest extends AnyFlatSpec with Matchers {
 
   def assertModule(input: String, result: AST): Assertion = {
     val parser = Parser()
-    val module = parser.run(new Reader(input))
+    val module = parser.run(new Reader(input), Nil)
```
Comment (mwu-tow) [inline common/syntax/specialization/shared/src/test/scala/org/enso/syntax/text/ParserTest.scala]: If we have overload that default to `Nil` idmap, then we shouldn't add it explicitly in all the calls. 
If we want to be explicit about losing ids, we should not have overload.

(I'd preferably have just a different name for an overload that doesn't take IDs, so it stays kind-of explicit but doesn't require repeating Nil.)

File: common/syntax/specialization/js/src/main/scala/org/enso/syntax/text/Main.scala [hunk 2]
```diff
@@ -19,4 +19,15 @@ object Parse {
       case e: Throwable => throw js.JavaScriptException(e.getMessage)
     }
   }
+  @JSExportTopLevel("parse_file")
+  def parse_file(content: String): (String, String) = {
```
Comment (mwu-tow) [inline common/syntax/specialization/js/src/main/scala/org/enso/syntax/text/Main.scala]: same naming concern as on ide

File: common/parser-service/src/main/scala/org/enso/parserservice/Protocol.scala [hunk 3]
```diff
@@ -12,11 +12,13 @@ import org.enso.syntax.text.Parser
...
   final case class ParseRequest(program: String, ids: Parser.IDMap)
-      extends Request
+    extends Request
+  final case class ParseFileRequest(content: String)
+    extends Request
 
   sealed trait Response
-  final case class Success(ast_json: String) extends Response
-  final case class Error(message: String)    extends Response
+  final case class Success(ast_json: String, metadata: String) extends Response
```
Comment (mwu-tow) [inline common/parser-service/src/main/scala/org/enso/parserservice/Protocol.scala]: Should this be types as `Metadata` or is it string with metadata serialized to JSON? Then it should get _json suffix. 
(please check if we can use richer type here and rely on auto-generated serialzietion)

File: common/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala [hunk 4]
```diff
@@ -155,7 +155,20 @@ class Parser {
   import Parser._
   private val engine = newEngine()
 
-  def run(input: Reader): AST.Module = run(input, Nil)
+  def run(fileContent: String): (AST.Module, Metadata) = {
+    fileContent.split("\n# [idmap]") match {
```
[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala]: Added constants. Generation can be added later when needed.

File: common/syntax/specialization/js/src/main/scala/org/enso/syntax/text/Main.scala [hunk 5]
```diff
@@ -19,4 +20,14 @@ object Parse {
       case e: Throwable => throw js.JavaScriptException(e.getMessage)
     }
   }
+  @JSExportTopLevel("parse_as_module")
+  def parse_as_module(program: String): String = {
```
Comment (mwu-tow) [inline common/syntax/specialization/js/src/main/scala/org/enso/syntax/text/Main.scala]: One more rename to `with_metadata` would be nice here.

File: common/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala [hunk 6]
```diff
@@ -151,11 +153,45 @@ class InternalError(reason: String, cause: Throwable = None.orNull)
...
+    Json.obj("ast" -> module.ast.toJson(), "metadata" -> module.metadata)
+}
+
+
 class Parser {
   import Parser._
   private val engine = newEngine()
 
-  def run(input: Reader): AST.Module = run(input, Nil)
+  def run_with_metadata(program: String): ModuleWithMetadata = {
```
Comment (mwu-tow) [inline common/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala]: please add doc comment here and on `run` explaining how they are different and how the metadata format looks like.

**Pull request merged**